### PR TITLE
Enable the RootCACertPublisher

### PR DESCRIPTION
**Pull request by jmprusi**
In order to get the ServiceAccounts working we will need to handle the `kube-root-ca.crt` configmap that contains the k8s CA certificate.

This PR wires the kube `rootCACert` controller into KCP.

Todo
- [x] Merge the PR that makes the RootCACertPublisher multi-workspace aware (https://github.com/kcp-dev/kubernetes/pull/49)
~~- [ ] Bump the Kubernetes version (https://github.com/kcp-dev/kcp/pull/671)~~

### CODE REVIEW

**Comment by jmprusi** [comment]
I'm aware that always syncing a configmap is not ideal, I will add another PR to make the scheduler a little bit more smart: don't sync a configmap like the kube-root-ca.crt if it's not really needed by a deployment or so.

**Comment by marun** [comment]
I get the value in syncing a configmap to a workload clusters, where it can be volume mounted into a pod. Is there also a use case for the configmap in kcp, where no pods can run? If not, maybe the configmap can be exposed by the syncer's virtual workspace only as required?

cc: @davidfestal

**Comment by marun** [comment]
Also, maybe a simple e2e to validate the creation of the configmap after namespace creation?

File: pkg/syncer/syncer.go [hunk 1]
```diff
@@ -176,7 +176,7 @@ func getAllGVRs(discoveryClient discovery.DiscoveryInterface, resourcesToSync ..
 			return nil, err
 		}
 	}
-	gvrstrs := []string{"namespaces.v1."} // A syncer should always watch namespaces
+	gvrstrs := []string{"namespaces.v1.", "configmaps.v1."} // A syncer should always watch namespaces, serviceaccounts and configmaps.
```
Comment (marun) [inline pkg/syncer/syncer.go:179]: Maybe enabling syncing of the configmap (and the associated testing) can be accomplished in a separate PR?

Comment (jmprusi) [reply pkg/syncer/syncer.go]: Done, will open a new pr for those changes with test.

File: pkg/server/controllers.go [hunk 2]
```diff
@@ -118,6 +120,48 @@ func (s *Server) installKubeNamespaceController(ctx context.Context, config *res
 	return nil
 }
 
+func (s *Server) installRootCAConfigMapController(ctx context.Context, config *rest.Config) error {
+	kubeClient, err := kubernetes.NewForConfig(config)
```
Comment (marun) [inline pkg/server/controllers.go:124]: Suggest adding a user agent.

```golang
config = rest.AddUserAgent(rest.CopyConfig(config), "root-ca-controller")
```

File: pkg/server/options/flags.go [hunk 3]
```diff
@@ -176,6 +176,9 @@ var (
 
 		// KCP Virtual Workspaces flags
 		"virtual-workspace-address", // Address of a stand-alone virtual workspace apiserver.
+
+		// Kubernetes ServiceAccount Controller
+		"root-ca-file", // If set, this root certificate authority will be included in service account's token secret. This must be a valid PEM-encoded CA bundle.
```
Comment (sttts) [inline pkg/server/options/flags.go:181]: is this required after the PR ?

File: pkg/server/controllers.go (hunk 2 above)
Comment (jmprusi) [reply pkg/server/controllers.go:125]: done!

File: pkg/server/options/flags.go (hunk 3 above)
Comment (jmprusi) [reply pkg/server/options/flags.go]: dropoed, thanks!

Comment (jmprusi) [reply pkg/server/options/flags.go:181]: turns out, that CI complains about it...  do you want me to disallow that flag?

Comment (marun) [reply pkg/server/options/flags.go:181]: Suggest rebasing on #681

Comment (jmprusi) [reply pkg/server/options/flags.go]: rebased.

File: pkg/server/controllers.go [hunk 4]
```diff
@@ -215,6 +217,49 @@ func (s *Server) installKubeServiceAccountTokenController(ctx context.Context, c
 	return nil
 }
 
+func (s *Server) installRootCAConfigMapController(ctx context.Context, config *rest.Config) error {
+	config = rest.AddUserAgent(rest.CopyConfig(config), "root-ca-configmap-controller")
```
Comment (sttts) [inline pkg/server/controllers.go:221]: nit: kube-root-ca-configmap-controller

File: pkg/server/controllers.go [hunk 5]
```diff
@@ -215,6 +217,49 @@ func (s *Server) installKubeServiceAccountTokenController(ctx context.Context, c
...
+
+	// TODO(jmprusi): We should make the CA loading dynamic when the file changes on disk.
+	caDataPath := s.options.Controllers.SAController.RootCAFile
+	if caDataPath == "" {
+		caDataPath = s.options.GenericControlPlane.SecureServing.SecureServingOptions.ServerCert.CertKey.CertFile
+	}
+
+	caData, err := os.ReadFile(caDataPath)
+	if err != nil {
+		return err
```
Comment (sttts) [inline pkg/server/controllers.go:235]: this need error context

File: pkg/server/controllers.go [hunk 6]
```diff
@@ -215,6 +217,49 @@ func (s *Server) installKubeServiceAccountTokenController(ctx context.Context, c
...
+	}
+
+	c, err := rootcacertpublisher.NewPublisher(
+		s.kubeSharedInformerFactory.Core().V1().ConfigMaps(),
+		s.kubeSharedInformerFactory.Core().V1().Namespaces(),
+		kubeClient,
+		caData,
+	)
+
```
Comment (sttts) [inline pkg/server/controllers.go:244]: no empty line

File: pkg/server/controllers.go [hunk 7]
```diff
@@ -215,6 +217,49 @@ func (s *Server) installKubeServiceAccountTokenController(ctx context.Context, c
...
+	if err != nil {
+		return err
+	}
+
+	s.AddPostStartHook("kcp-start-kube-rootcacertpublisher-controller", func(hookContext genericapiserver.PostStartHookContext) error {
```
Comment (sttts) [inline pkg/server/controllers.go:249]: use same controller name as for the useragent

File: pkg/server/controllers.go [hunk 8]
```diff
@@ -215,6 +217,50 @@ func (s *Server) installKubeServiceAccountTokenController(ctx context.Context, c
...
+		s.kubeSharedInformerFactory.Core().V1().ConfigMaps(),
+		s.kubeSharedInformerFactory.Core().V1().Namespaces(),
+		kubeClient,
+		caData,
+	)
+	if err != nil {
+		return fmt.Errorf("error creating %s controller: %w", rootCAConfigMapControllerName, err)
+	}
+
+	s.AddPostStartHook(rootCAConfigMapControllerName, func(hookContext genericapiserver.PostStartHookContext) error {
```
Comment (sttts) [inline pkg/server/controllers.go:250]: see other example for how this is called

File: test/e2e/authorizer/rootcacertconfigmap_test.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "e2e-sa-",
+		},
+	}, metav1.CreateOptions{})
+	require.NoError(t, err, "failed to create namespace")
+
+	t.Log("Waiting for default configmap to be created")
+	require.Eventually(t, func() bool {
+		configmap, err := kubeClient.CoreV1().ConfigMaps(namespace.Name).Get(ctx, DefaultRootCACertConfigmap, metav1.GetOptions{})
+
```
Comment (sttts) [inline test/e2e/authorizer/rootcacertconfigmap_test.go:64]: no empty line

File: test/e2e/authorizer/rootcacertconfigmap_test.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+		if apierrors.IsNotFound(err) {
+			return false
+		} else if err != nil {
```
Comment (sttts) [inline test/e2e/authorizer/rootcacertconfigmap_test.go:67]: no need for the else if

**Review by sttts — APPROVED** [review]
